Fiddling with datum ID assignment

esoergel opened the pull request:
## Technical Summary
This is largely a series of refactors to make the code a bit easier to understand.  There are still some error cases that are demonstrated in the tests, I'll comment there in-line.  Mostly opening this PR to get feedback on the refactors and to solicit ideas on how to proceed from here.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
This stuff is extensively tested, and the behavior changes I made are pretty minimal, but I'd definitely appreciate a gut check on safety from someone else versed in this code.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

orangejenny committed:
Fixed another bit of weird multi-select datum behavior

esoergel committed:
it is now

esoergel committed:
Move utility functions outside of the method

esoergel committed:
remove trailing underscore now there's no conflict

esoergel committed:
Make code read a little easier

esoergel committed:
Move utilities back in line and refactor

esoergel committed:
Update the refs when they're changed, not later

esoergel committed:
This was inadvertently the same as the test before

esoergel committed:
Add a couple more test cases

esoergel committed:
compare node name, not datum type

esoergel committed:
rename current and parent to previous and current

The "parent" module is the one being added to the chain, so it seems
mroe sensible to call that one "current"

esoergel committed:
Use "selected_cases" for multiselect modules

This is a deliberate behavior changes
`selected_cases` will be used only for multiselect modules themselves,
and it will always be used for multiselect modules, regardless of the
parent select mode

esoergel committed:
Check the full datum, not just ID

We're checking whether the datums at the same position match.  ID isn't
enough to do that conclusively

esoergel committed:
Return a new list rather than mutating one

The test changes here are illustrative of the reason why (besides
clarity).  We inserted into a list, mutating it, while iterating over
that same list.  This is why the `list()` coercion in the `for`
statement was necessary - it made a copy of the old version of the list.
However, each time we add new elements to the current position, the `i`
variable drifts behind.  This doesn't affect anything where there's only
one insertion, but these two tests called that insertion twice, so the
second was off-by-one.

esoergel committed:
Update test comments

esoergel commented on corehq/apps/app_manager/tests/test_suite_multi_select_case_list.py:
```diff
@@ -352,28 +350,25 @@ def test_parent_selects_parent_same_type(self):
         ])
         self.assert_module_datums(self.m1.id, [
-            ('datum', 'case_id_beneficiary'),
+            ('datum', 'case_id_beneficiary'),  # From m1, but copied from m2
```
As a refresher:
```
# All of these tests use the same app structure:
# m0 is a parent module and its case list is multi-select
# m1 is a child of m0 and uses the same case type, this is where the tests focus
# m2 is a standalone module of the same case type
# m3 is a standalone module of another case type
```
The issue in this test is that `get_select_chain_meta` only returns the select chain for m1, since it doesn't use `parent_select`, but then in `add_parent_datums` we compare that chain against that of its parent, which contains `case_id_beneficiary` from m2 and `selected_cases` from m0.

When these two are aligned, m1's `case_id` is improperly correlated with `m2`'s `case_id_beneficiary` and reassigned that name, even though both cases come from different case lists.  `get_add_case_preloads_case_id_xpath` is pretty naive, and just anticipates `case_id`.

esoergel commented on corehq/apps/app_manager/tests/test_suite_multi_select_case_list.py:
```diff
@@ -398,12 +392,34 @@ def test_select_parent_that_selects_other_different_case_type(self):
+            ('datum', 'case_id_beneficiary'),
+            ('datum', 'case_id')
+        ])
+        # This is wrong - get_add_case_preloads_case_id_xpath expects the child
+        # module datum to be renamed because of a conflict with the parent, but
+        # in this case the parent (m0) doesn't conflict, as it's a multiselect.
+        self.assert_form_datums(self.m1f0, 'case_id_beneficiary')
```
The comment describes the issue here.  It's especially confusing, as the parent module's datum doesn't get pulled in to the suite.

snopoke commented:
@esoergel I'm going to put some time aside tomorrow (May 13) to look at this.

snopoke commented on corehq/apps/app_manager/suite_xml/sections/entries.py:
```diff
@@ -879,24 +876,26 @@ def add_parent_datums(self, datums, module):
+                        set_id(datum.datum, '_'.join((datum.datum.id, datum.case_type)))
+
+                    set_id(this_datum_meta.datum, parent_datum_meta.datum.id)
+            if this_datum_meta:
+                ret.append(this_datum_meta)
```
I think this is incorrect as it will interleave the parent module datums with the child modules. I thought the 'datums' for the child module could all be added after the loop:

```
ret.extend(datums)
```

But that also breaks other tests. I think this whole method maybe needs to be re-structured to not 'zip' the datums (or we just leave it to mutate the list).

snopoke committed:
only render the suite once

esoergel commented on corehq/apps/app_manager/tests/test_suite_multi_select_case_list.py:
```diff
@@ -297,14 +298,16 @@ def setUp(self):
         self.m3, m3f0 = self.factory.new_basic_module('m3', self.OTHER_CASE_TYPE)
         m3f0.requires = 'case'
 
+        self._render_suite.reset_cache(self)
```
:+1: thanks, I was planning on doing something to that effect, this is a clean approach

snopoke committed:
get case session var from entry datums

snopoke committed:
create new helper functions

snopoke committed:
fix datum ordering

snopoke committed:
move assert to test class mixin

snopoke committed:
make SuiteMixin extend TestXmlMixin

snopoke committed:
remove unncessary mixins

snopoke committed:
Adjust datum ID for child module with different case type

snopoke committed:
short cut to datum_meta.datum.id

orangejenny reviewed (commented):
This looks good to me pending resolving Simon's comment [here](https://github.com/dimagi/commcare-hq/pull/31613/commits/ae23b1a153c1e0d9f6bfc25e1cf258f3673b1077?w=1#r871501383).

snopoke committed:
check for None

snopoke commented:
I pulled out my commits and added a few more here: https://github.com/dimagi/commcare-hq/pull/31623

snopoke committed:
use datums for session case ID if root module OR parent select

snopoke committed:
test for child module, same case type, no parent select

esoergel committed:
Merge pull request #31623 from dimagi/sk/datums

some more app XML tweaks and fixes

esoergel commented on corehq/apps/app_manager/suite_xml/sections/entries.py:
```diff
@@ -957,44 +953,44 @@ def _get_detail_from_module(self, module, detail_type):
+        if datum.action:
+            if hasattr(datum.action, 'case_indices'):
+                # This is an advanced module
+                for case_index in datum.action.case_indices:
```
That's tricky actually, since we need the `else` part of this block if `action` isn't set.  It could be
```python
for case_index in getattr(datum.action, 'case_indices', [CaseIndex(tag='basic')]):
```
But that also seems kinda messy

esoergel committed:
PR feedback rename

esoergel committed:
Don't zip datum lists together

Hopefully this is simpler iteration

esoergel committed:
Simplify ID conflict handling

This does two things:

Looks up the case tag from the datum that's being modified, rather than the
current this_datum_meta.  This makes more sense to me, and doesn't break any tests.

It moves this logic outside of the if statement, so conflicting parents will
always be corrected.  This also doesn't break any tests.

esoergel committed:
test comments

esoergel committed:
Don't overwrite import

esoergel commented:
Well this was a marathon.  I think this is in a good place now, at least I'm done fiddling with it.  Thanks a ton Jenny and Simon for your help.

snopoke commented on corehq/apps/app_manager/suite_xml/sections/entries.py:
```diff
@@ -1042,3 +981,54 @@ def _get_detail_from_module(self, module, detail_type):
+            and this_datum_meta.datum.ROOT_NAME == parent_datum_meta.datum.ROOT_NAME)
+
+
+def _same_id(this_datum_meta, parent_datum_meta):
```
I think this function is now unused

snopoke commented on corehq/apps/app_manager/suite_xml/sections/entries.py:
```diff
@@ -844,150 +869,64 @@ def get_manual_datum(action_, parent_filter_=''):
+        ret = []
+        datums_remaining = list(datums)
+        for parent_datum_meta in parent_datums:
+            if parent_datum_meta.id in datum_ids:
+                datum = datum_ids[parent_datum_meta.id]
+                set_id(datum, f'{datum.id}_{datum.case_type}')
```
I expected this change to break the `BasicModuleAsChildTest.test_child_module_parent_no_parent_select_same_case_type` test. Debugging locally I see that this first renames `case_id` to `case_id_gold_fish` and then on line 926 it renames it back again which is why the test doesn't break. 

I'm not sure how I feel about that but perhaps just a comment would be sufficient:
```suggestion
            if parent_datum_meta.id in datum_ids:
                # We assume that a conflict is unwanted and rename. This may get undone below.
                datum = datum_ids[parent_datum_meta.id]
                set_id(datum, f'{datum.id}_{datum.case_type}')
```

esoergel replied:
I agree - I'm not crazy about it as-is.  I just tried a few variations, but wasn't able to get anything that looked better and still worked.  It'd be nice to have this only iterate over datum IDs remaining - it's weird that it could in theory mutate datums that have already been added to `ret`.  I think my ideal would be to have this pulled out of the loop and have a separate block that checks for ID conflicts when adding the remaining datums.  Anyways, I'm punting for now and just adding that comment.

snopoke reviewed (approved):
small comments but otherwise LGTM

esoergel committed:
PR feedback

esoergel committed:
Minor simplifications

I think this shows intent a little better - if they represent the same
thing, make sure the IDs match, and add it
